src/mpi/romio/adio/common/utils.c

```diff
@@ -6,6 +6,47 @@
 #include <adio.h>
 #include <limits.h>
 
+int ADIOI_Type_get_combiner(MPI_Datatype datatype, int *combiner)
+{
+    int ret;
+#if MPI_VERSION >= 4
+    MPI_Count ni, na, nc, nt;
+    ret = MPI_Type_get_envelope_c(datatype, &ni, &na, &nc, &nt, combiner);
+#else
+    int ni, na, nt;
+    ret = MPI_Type_get_envelope(datatype, &ni, &na, &nt, combiner);
+#endif
+    return ret;
+}
+
+int ADIOI_Type_ispredef(MPI_Datatype datatype, int *flag)
+{
+    int ret, combiner = 0;
+    ret = ADIOI_Type_get_combiner(datatype, &combiner);
+    switch (combiner) {
+        case MPI_COMBINER_NAMED:
+        case MPI_COMBINER_F90_INTEGER:
+        case MPI_COMBINER_F90_REAL:
+        case MPI_COMBINER_F90_COMPLEX:
+            *flag = 1;
+            break;
+        default:
+            *flag = 0;
+            break;
+    }
+    return ret;
+}
+
+int ADIOI_Type_free(MPI_Datatype * datatype)
```

Comment:
How about rename it to `ADIOI_Type_free_ifnotpredef`?

Reply:
Again, I was following the pattern of `ADIOI_Datatype_iscontig()`. I'm used to write predicate routines that check for the "positive" property, not the opposite (`is_<something>` rather than `is_not_<something>`), although I admit that is ultimately a matter of style and personal preference. 

If you want to change it anyway, I'll surrender. But please look the code around in ROMIO and confirm whether your  choice follows the established patterns elsewhere.

Reply:
I just want to reflect the semantics in the function name. The semantics of the routine is "free the datatype if it is not a predefined type", right? Just `ADIOI_Type_free` makes people wonder.

Reply:
> Just ADIOI_Type_free makes people wonder.

Oh. Sorry, I totally misunderstood your comment. I misread, I thought you were talking about `ADIOI_Type_ispredef`!

I believe you are right, `ADIOI_Type_free` is not explicit enough regarding its intention. I don't like your naming 100%, but I cannot come up with anything better, so if you still want it, I can implement your suggestion. Should I?

Reply:
I certainly would like the name be more explicit.

Reply:
I renamed it to `MPI_Type_dispose()` and added a comment explaining its semantics. If you do not like my new name, then I'll give up trying to be inventive, and I'll just go your way.